src/transaction/atomic_replication_helper.cpp

```diff
@@ -301,162 +354,410 @@ namespace cublog
+      {
+	--first_work_log_it;
+      }
+    assert (first_work_log_it->is_control ());
+    ++first_work_log_it; // up one to the first actual work
+    assert (!first_work_log_it->is_control ());
+
+    // debug check, all entries before the first non-control must be control entries
+    assert (std::all_of (m_log_vec.cbegin (), first_work_log_it,
+			 [] (const atomic_log_entry &entry)
+    {
+      return entry.is_control ();
+    }));
+
+    for (auto apply_it = first_work_log_it; apply_it != m_log_vec.end (); ++apply_it)
+      {
+	const atomic_log_entry &log_entry = *apply_it;
+	assert (!log_entry.is_control ());
+	log_entry.apply_log_redo (thread_p, m_redo_context);
+	m_page_ptr_bookkeeping.unfix_page (thread_p, log_entry.m_vpid);
+      }
+
+    m_log_vec.erase (first_work_log_it, m_log_vec.end ());
+
+    if (prm_get_bool_value (PRM_ID_ER_LOG_PTS_ATOMIC_REPL_DEBUG))
+      {
+	dump ("sequence::apply_and_unfix END");
+      }
+
+    assert (all_log_entries_are_control ());
   }
 
-  void
-  atomic_replication_helper::atomic_log_sequence::start_postpone_sequence ()
+  LOG_LSA
+  atomic_replication_helper::atomic_log_sequence::get_start_lsa () const
   {
-    assert (m_is_sysop);
-    assert (!m_postpone_started);
-
-    m_postpone_started = true;
+    // TODO: start LSA can also be obtained from the first log entry
+    return m_start_lsa;
   }
 
-  bool
-  atomic_replication_helper::atomic_log_sequence::is_postpone_sequence_started () const
+  void
+  atomic_replication_helper::atomic_log_sequence::append_control_log (LOG_RECTYPE rectype, LOG_LSA lsa)
   {
-    //assert ((m_postpone_started && m_is_sysop) || !m_is_sysop);
+    m_log_vec.emplace_back (lsa, rectype);
 
-    return m_postpone_started;
+    if (prm_get_bool_value (PRM_ID_ER_LOG_PTS_ATOMIC_REPL_DEBUG))
+      {
+	dump ("sequence::append_control_log END");
+      }
   }
 
   void
-  atomic_replication_helper::atomic_log_sequence::complete_one_postpone_sequence ()
+  atomic_replication_helper::atomic_log_sequence::append_control_log_sysop_end (
+	  LOG_LSA lsa, LOG_SYSOP_END_TYPE sysop_end_type, LOG_LSA sysop_end_last_parent_lsa)
   {
-    assert (m_is_sysop);
-    assert (m_postpone_started);
+    m_log_vec.emplace_back (lsa, sysop_end_type, sysop_end_last_parent_lsa);
 
-    ++m_end_pospone_count;
+    if (prm_get_bool_value (PRM_ID_ER_LOG_PTS_ATOMIC_REPL_DEBUG))
+      {
+	dump ("sequence::append_control_log_sysop_end END");
+      }
   }
 
   bool
-  atomic_replication_helper::atomic_log_sequence::is_at_least_one_postpone_sequence_completed () const
+  atomic_replication_helper::atomic_log_sequence::all_log_entries_are_control () const
   {
-    if (m_end_pospone_count > 0)
+    return std::all_of (m_log_vec.cbegin (), m_log_vec.cend (),
+			[] (const atomic_log_entry &entry)
+    {
+      return entry.is_control ();
+    });
+  }
+
+  bool
+  atomic_replication_helper::atomic_log_sequence::can_purge ()
+  {
+    assert (all_log_entries_are_control ());
```

Comment:
Is this assertion satisfied because it is only called after apply_and_unfix is ​​executed? 

LOG_ATOMIC_SYSOP (1) -> non control log -> LOG_ATOMIC_SYSOP without log  record (2) -> non control log ->  LOG_SYSOP_END (3) -> LOG_SYSOP_END (4) 

When the control log record of `(3) LOG_SYSOP_END` is replicated,
(3) LOG_SYSOP_END is a log record corresponding to LOG_ATOMIC_SYSOP (2). 
In this case, the log records between (1) and (3) will be applied even if the (1) is not ended with commit.

Reply:
Yes, this is the intention.
See previous comments and also comments in function `can_purge`.
The direct reason is explained in https://github.com/CUBRID/cubrid/pull/3910#discussion_r1012332005 - same page can be modified by multiple levels of a nested atomic sequence.

Although, on a second thought, there might be a better way to handle this. Given that we can take advantage of the `page_ptr_bookkeeping` functionality which reference-counts fixes, thus allowing the both application of nested sequences in same pages and also consistent unfix.

I will add another issue to re-re-refactor this. :) It is maybe lower prio.

Reply:
Added http://jira.cubrid.org/browse/LETS-585 to address both this issue and the issue of "true atomic replication for the entire atomic sequence and sub-sequences".